src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/RegexIgnoreCaseTests.cs

```diff
@@ -0,0 +1,79 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Globalization;
+using System.Tests;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.Text.RegularExpressions.Tests
+{
+    // This test takes a long time to run since it needs to compute all possible lowercase mappings across
+    // 3 different culutres and then creates Regex matches for all of our engines for each mapping.
+    [OuterLoop]
+    public class RegexIgnoreCaseTests
+    {
+        [Theory]
+        [MemberData(nameof(Unicode_IgnoreCase_TestData))]
+        public async Task Unicode_IgnoreCase_Tests(RegexEngine engine, string culture, string pattern, string input, RegexOptions options)
+        {
+            if ((options & RegexOptions.CultureInvariant) == 0)
+            {
+                using var _ = new ThreadCultureChange(culture);
+                await ValidateMatch(pattern, input);
+                await ValidateMatch(input, pattern);
+            }
+            else
+            {
+                await ValidateMatch(pattern, input);
+                await ValidateMatch(input, pattern);
+            }
+
+            async Task ValidateMatch(string pattern, string input)
+            {
+                Regex regex = await RegexHelpers.GetRegexAsync(engine, pattern, options | RegexOptions.IgnoreCase);
+                Assert.True(regex.IsMatch(input));
+            }
+        }
+
+        public static IEnumerable<object[]> Unicode_IgnoreCase_TestData()
+        {
+            foreach ((string culture, char pattern, char input) in GetCaseEquivalencesPerCulture())
+            {
+                foreach (RegexEngine engine in RegexHelpers.AvailableEngines)
+                {
+                    // Source generated engine does not support setting different culutres yet.
+                    if (engine == RegexEngine.SourceGenerated && !string.IsNullOrEmpty(culture))
+                        continue;
+
+                    yield return new object[] { engine, culture, $"{pattern}", $"{input}", RegexOptions.None};
+                    if (string.IsNullOrEmpty(culture))
+                    {
+                        // For the Invariant culture equivalences also test to get the same behavior with RegexOptions.CultureInvariant.
+                        yield return new object[] { engine, culture, $"{pattern}", $"{input}", RegexOptions.CultureInvariant };
+                    }
+                }
+            }
+        }
+
+        public static IEnumerable<(string, char, char)> GetCaseEquivalencesPerCulture()
+        {
+            foreach (string cultureName in new[] { "", "en-US", "tr-TR" })
```

Comment:
If .NET adds support for another culture that requires a "special case", will any test fail? I mean: in principle, we could get this test to enumerate over all supported cultures, and it would then fail if a new one was added that needed attention. But I'm guessing that's way too slow.

Reply:
yeah we would need to enumerate all cultures in order to detect such cases and break accordingly. FWIW, In order to come up with the different behaviors we support with this PR I did run all permutations across all enumerated cultures, and found that there are only three distinct behaviors amongst the 870 that .NET supports, and all of those only differ on how they handle characters: 'i', 'I', '\u0130', '\u0131'. Other than that, all cultures behave exactly the same when used like: `culture.TextInfo.ToLower(c)`

Reply:
Yep, I'm sure of that - I know Steve did the same test - just thinking of the hypothetical future where we add another. not sure it's worth "protecting against" though.